src/realm/sync/client.cpp

```diff
@@ -1784,47 +1852,90 @@ void SessionWrapper::on_connection_state_changed(ConnectionState state,
     }
 }
 
+void SessionWrapper::init_progress_handler()
+{
+    uint64_t unused = 0;
+    ClientHistory::get_upload_download_bytes(m_db.get(), m_reported_progress.final_downloaded, unused,
+                                             m_reported_progress.final_uploaded, unused, unused);
+}
 
-void SessionWrapper::report_progress(bool only_if_new_uploadable_data)
+void SessionWrapper::report_progress(bool is_download, bool is_completed, bool only_if_new_uploadable_data)
```

Comment:
I find the logic in this function hard to follow (with a lot of checks). I will put some thought into a cleaner approach.

Reply:
i also want to make it cleaner with #7451 Reusing the same method for clear intent in this change was simpler initially, but became pretty ugly as you've noticed. I've almost finished the next change splitting this function for download and upload, so hopefully it'll get better.